engine/runtime/src/main/java/org/enso/compiler/pass/analyse/PrivateModuleAnalysis.java

```diff
@@ -0,0 +1,171 @@
+package org.enso.compiler.pass.analyse;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import org.enso.compiler.context.InlineContext;
+import org.enso.compiler.context.ModuleContext;
+import org.enso.compiler.core.IR;
+import org.enso.compiler.core.ir.Expression;
+import org.enso.compiler.core.ir.Module;
+import org.enso.compiler.core.ir.expression.errors.ImportExport;
+import org.enso.compiler.core.ir.module.scope.Export;
+import org.enso.compiler.core.ir.module.scope.Import;
+import org.enso.compiler.data.BindingsMap;
+import org.enso.compiler.pass.IRPass;
+import org.enso.interpreter.util.ScalaConversions;
+import org.enso.pkg.QualifiedName;
+import scala.Option;
+import scala.collection.immutable.Seq;
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * Iterates through all the imports and exports of non-synthetic modules and ensures that:
+ * <ul>
+ *   <li>No private module is exported</li>
+ *   <li>No private module from a different project is imported</li>
+ * </ul>
+ * Inserts errors into imports/exports IRs if the above conditions are violated.
+ */
+public class PrivateModuleAnalysis implements IRPass {
```

Comment:
I'd made the class `final` myself. But yes, it is duplicate check over `private` constructor.

Reply:
Made final in 42fe69b384c3ceaed518564a15618c816d0974cf